cmd/commandfuncs.go

```diff
@@ -640,17 +642,18 @@ commands:
 	return caddy.ExitCodeSuccess, nil
 }
 
-// apiRequest makes an API request to the endpoint adminAddr with the
+// ApiRequest makes an API request to the endpoint adminAddr with the
 // given HTTP method and request URI. If body is non-nil, it will be
-// assumed to be Content-Type application/json.
-func apiRequest(adminAddr, method, uri string, headers http.Header, body io.Reader) error {
+// assumed to be Content-Type application/json. The caller should
+// close the response body.
+func ApiRequest(adminAddr, method, uri string, headers http.Header, body io.Reader) (*http.Response, error) {
```

Comment:
```suggestion
func APIRequest(adminAddr, method, uri string, headers http.Header, body io.Reader) (*http.Response, error) {
```

As mentioned in another comment, I have some concerns with this signature, especially since we're going to be exporting it. I get a lot of anxiety exporting functions and types these days 😅

Can the function take input to determine the admin address itself instead?

Or maybe can we leave this function as-is (except for, I guess, returning the response) and create a new function that takes the necessary inputs for determining the admin address. It would be realluy good if we didn't have to repeat this logic and have callers do that.

I also wonder if an exported function here could be named a little more specifically, like `AdminAPIRequest()` or something.

Reply:
Okay, this is what I did:

- Renamed it to `AdminAPIRequest()`
- Signature stayed the same, because the options are either
   1. Use `Flags` as the first arg instead and try to automatically parse the command's flags to get the address; this makes the Flags leak outside of the command's function itself, so this is kinda bleh (I did try it in code to see if it made sense)
   2. Change `adminAddr` to three string inputs instead, i.e. address, config, adapter... but that's also gross (or some setup where a struct with those strings are passed in, also gross)
   3. Keep it the same
- Moved the duplicated logic for determining the admin address to a helper `caddycmd.DetermineAdminAPIAddress()` which the 3 common CLI flags, now `run`, `stop`, `trust` and `untrust` all use this
- Moved the code that performs the admin API call and decodes the response in `trust` and `untrust` to a `caddypki.rootCertFromAdmin()` to avoid duplication.

This seems cleaner, but it doesn't _directly_ address your concerns. Not sure what the other options are, if there are any.

Reply:
Thanks - of the all-not-great options, I like your approach the best I guess :)